### PR TITLE
Adjust text width constraint based on available space

### PR DESCRIPTION
**Pull request by ales-erjavec**

### CODE REVIEW

**Comment by ales-erjavec** [comment]
ref: biolab/orange3#1768

File: Orange/canvas/canvas/items/linkitem.py [hunk 1]
```diff
@@ -533,10 +538,35 @@ def __updateText(self):
...
+        if available_width > 0:
+            # Get the ideal text width if it was unconstrained
+            doc = self.linkTextItem.document().clone(self)
+            doc.setTextWidth(-1)
+            idealwidth = doc.idealWidth()
+            doc.deleteLater()
+
+            # Constrain the text width  but not below a certain min width
+            minwidth = 100
+            textwidth = max(minwidth, min(available_width, idealwidth))
```
Comment (lanzagar) [inline Orange/canvas/canvas/items/linkitem.py:558]: Why is idealwidth needed here? I tried with just `max(minwidth, available_width)` and it seemed to behave the same.

File: Orange/canvas/canvas/items/linkitem.py [hunk 2]
```diff
@@ -533,10 +538,35 @@ def __updateText(self):
...
+        # between the two ends
+        if not path.isEmpty():
+            # Use the distance between the start/end points as a measure of
+            # available space
+            diff = path.pointAtPercent(0.0) - path.pointAtPercent(1.0)
+            available_width = math.sqrt(diff.x() ** 2 + diff.y() ** 2)
+        else:
+            available_width = -1
+
+        if available_width > 0:
```
Comment (lanzagar) [inline Orange/canvas/canvas/items/linkitem.py]: Why is this in a separate `if`? Couldn't it be merged with the `if` above?

**Review by lanzagar — COMMENTED** [review]
I think this solution is the best so far. It can still result in some awkward breaks, but the dynamic changes make it less of a problem and just make for a better UI.

File: Orange/canvas/canvas/items/linkitem.py (hunk 1 above)
Comment (ales-erjavec) [reply Orange/canvas/canvas/items/linkitem.py:558]: Try running orange with:
```
QT_DRAW_SCENE_ITEM_RECTS=1 orange-canvas
```
to see the difference (the text label's bounding box would grow to the width/length of the entire link)

File: Orange/canvas/canvas/items/linkitem.py (hunk 2 above)
Comment (ales-erjavec) [reply Orange/canvas/canvas/items/linkitem.py]: It was a part of a longer chain of conditional logic. Removed now.

File: Orange/canvas/canvas/items/linkitem.py (hunk 1 above)
Comment (lanzagar) [reply Orange/canvas/canvas/items/linkitem.py:558]: I imagined that would be the case. But is that a problem? Unless running with visible rects, I can't see a negative effect of simply using the available space (it is already used by the link too).

The 5 extra lines of code are also not that big of an issue, I am just curious why we need them.

**Pull request merged**